src/libcrun/handlers/spin.c

```diff
@@ -0,0 +1,116 @@
+/*
+ * crun - OCI runtime written in C
+ *
+ * Copyright (C) 2017, 2018, 2019, 2020, 2021 Giuseppe Scrivano <giuseppe@scrivano.org>
+ * crun is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * crun is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with crun.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#define _GNU_SOURCE
+
+#include <config.h>
+#include "../custom-handler.h"
+#include "../linux.h"
+#include <sys/stat.h>
+
+#ifdef HAVE_DLOPEN
+#  include <dlfcn.h>
+#endif
+
+#ifdef HAVE_SPIN
+
+#endif
+
+#if HAVE_DLOPEN && HAVE_SPIN
+static int
+spin_exec (void *cookie arg_unused, libcrun_container_t *container arg_unused,
+           const char *pathname arg_unused, char *const argv[] arg_unused)
+{
+  // wasmtime fails to determine default config path if $HOME is not set
+  char *newenviron[] = { "HOME=/root", NULL };
+  char *newargv[] = { "/bin/spin", "up", "--listen", "--listen", NULL };
+
+  // spin up needs a /tmp folder
+  struct stat st = { 0 };
+  if (stat ("/tmp", &st) == -1)
+    {
+      mkdir ("/tmp", 0700);
+    }
+
+  execve (newargv[0], newargv, newenviron);
+  perror ("execve");
+  exit (EXIT_FAILURE);
+}
+
+static int
+spin_load (void **cookie, libcrun_error_t *err)
```

Comment:
Non-blocking, not-urgent: possibly `spin_load` can be used to check if spin binary exists on host or not.

Reply:
I think it is a good thing to have an explicit error message instead of "file not found in path". So I added the check if the file exists on the host filesystem. Thanks for the hint!